File: low-level-design/lld-coding/multi-language/04-page-view-counter/page_view_counter.js
/*
 * JavaScript: Page View Counter - Race Condition Demo
 * ====================================================
 * Shows:
 * 1. Why single-threaded JS usually doesn't have races
 * 2. Worker threads CAN cause races
 * 3. Using Atomics for shared memory
 * 4. Why async/await doesn't need locks
 *
 * Note: Requires Node.js 12+ for Worker threads
 */

const { Worker, isMainThread, parentPort, workerData } = require('worker_threads');

console.log("================================================================================");
console.log("          JAVASCRIPT: RACE CONDITION DEMONSTRATION");
console.log("================================================================================");

// ============================================================================
// Version 1: Single-threaded (NO RACE - event loop is sequential)
// ============================================================================
class PageViewCounterSingleThreaded {
    constructor() {
        this.views = 0;
    }

    increment() {
        // Even with async operations, no race in single thread
        const temp = this.views;
        this.views = temp + 1;
    }

    getViews() {
        return this.views;
    }
}

async function demonstrateSingleThreaded() {
    console.log("\n1. SINGLE-THREADED VERSION (No Race Possible)");
    console.log("--------------------------------------------------------------------------------");
    console.log("JavaScript's event loop is single-threaded, so no race conditions by default");

    const counter = new PageViewCounterSingleThreaded();

    // Even with 1000 async operations, they run sequentially
    const promises = [];
    for (let i = 0; i < 1000; i++) {
        promises.push(
            new Promise(resolve => {
                counter.increment();
                resolve();
            })
        );
    }

    await Promise.all(promises);

    console.log(`Expected: 1000`);
    console.log(`Actual:   ${counter.getViews()}`);
    console.log(`Status:   ✅ Correct! (Event loop is sequential)`);
}

// ============================================================================
// Version 2: Worker Threads WITHOUT synchronization (BROKEN)
// ============================================================================
function demonstrateWorkerThreadsBroken() {
    return new Promise((resolve) => {
        console.log("\n2. WORKER THREADS - BROKEN (No Synchronization)");
        console.log("--------------------------------------------------------------------------------");
        console.log("With SharedArrayBuffer, worker threads CAN have race conditions!");

        // Shared memory between workers
        const sharedBuffer = new SharedArrayBuffer(4); // 4 bytes = 1 int32
        const sharedArray = new Int32Array(sharedBuffer);
        sharedArray[0] = 0; // Initial value

        let workersCompleted = 0;
        const totalWorkers = 10;

        // Create 10 workers, each incrementing 100 times
        for (let i = 0; i < totalWorkers; i++) {
            const worker = new Worker(__filename, {
                workerData: { sharedBuffer, iterations: 100, useAtomic: false }
            });

            worker.on('exit', () => {
                workersCompleted++;
                if (workersCompleted === totalWorkers) {
                    console.log(`Expected: 1000 (10 workers × 100 increments)`);
                    console.log(`Actual:   ${sharedArray[0]}`);
                    console.log(`Lost:     ${1000 - sharedArray[0]} increments`);
                    console.log(`Status:   ${sharedArray[0] === 1000 ? '✅ Correct' : '❌ RACE CONDITION DETECTED!'}`);
                    resolve();
                }
            });
        }
    });
}

// ============================================================================
// Version 3: Worker Threads WITH Atomics (FIXED)
// ============================================================================
function demonstrateWorkerThreadsFixed() {
    return new Promise((resolve) => {
        console.log("\n3. WORKER THREADS - FIXED (With Atomics)");
        console.log("--------------------------------------------------------------------------------");
        console.log("Atomics.add() provides lock-free synchronization");

        const sharedBuffer = new SharedArrayBuffer(4);
        const sharedArray = new Int32Array(sharedBuffer);
        sharedArray[0] = 0;

        let workersCompleted = 0;
        const totalWorkers = 10;

        for (let i = 0; i < totalWorkers; i++) {
            const worker = new Worker(__filename, {
                workerData: { sharedBuffer, iterations: 100, useAtomic: true }
            });

            worker.on('exit', () => {
                workersCompleted++;
                if (workersCompleted === totalWorkers) {
                    console.log(`Expected: 1000`);
                    console.log(`Actual:   ${sharedArray[0]}`);
                    console.log(`Status:   ✅ Correct! Atomics prevent race conditions`);
                    resolve();
                }
            });
        }
    });
}

// ============================================================================
// Version 4: Async/Await Pattern (NO RACE - single threaded)
// ============================================================================
class PageViewCounterAsync {
    constructor() {
        this.views = 0;
    }

    async increment() {
        // Even with delays, no race - event loop is sequential
        await new Promise(resolve => setTimeout(resolve, 1));
        this.views++;
    }

    getViews() {
        return this.views;
    }
}

async function demonstrateAsyncAwait() {
    console.log("\n4. ASYNC/AWAIT PATTERN (No Race - Sequential Event Loop)");
    console.log("--------------------------------------------------------------------------------");
    console.log("async/await doesn't create threads - still single-threaded!");

    const counter = new PageViewCounterAsync();

    // Even with async/await, operations are sequential
    const promises = [];
    for (let i = 0; i < 100; i++) {
        promises.push(counter.increment());
    }

    await Promise.all(promises);

    console.log(`Expected: 100`);
    console.log(`Actual:   ${counter.getViews()}`);
    console.log(`Status:   ✅ Correct! async/await is still single-threaded`);
}

// ============================================================================
// Worker Thread Code
// ============================================================================
if (!isMainThread) {
    const { sharedBuffer, iterations, useAtomic } = workerData;
    const sharedArray = new Int32Array(sharedBuffer);

    for (let i = 0; i < iterations; i++) {
        if (useAtomic) {
            // CORRECT: Atomic increment (thread-safe)
            Atomics.add(sharedArray, 0, 1);
        } else {
            // WRONG: Non-atomic increment (race condition)
            const temp = sharedArray[0];
            sharedArray[0] = temp + 1;
        }
    }

    process.exit(0);
}

// ============================================================================
// Main Execution
// ============================================================================
async function main() {
    if (isMainThread) {
        // Run all demonstrations
        await demonstrateSingleThreaded();
        await demonstrateWorkerThreadsBroken();
        await demonstrateWorkerThreadsFixed();
        await demonstrateAsyncAwait();
        printSummary();
    }
}

function printSummary() {
    console.log("\n================================================================================");
    console.log("                          SUMMARY");
    console.log("================================================================================");
    console.log(`
🎯 Key Lessons:

1. JavaScript is SINGLE-THREADED by default
   - Event loop processes one task at a time
   - No race conditions in normal async code
   - async/await is NOT multi-threading

2. Worker Threads CAN have races
   - SharedArrayBuffer allows shared memory
   - Multiple workers = true parallelism
   - Need synchronization like other languages

3. Synchronization in JavaScript:
   a) Atomics API
      - Atomics.add(), Atomics.sub(), etc.
      - Lock-free operations
      - Only works with SharedArrayBuffer

   b) Message passing (recommended)
      - postMessage() / onmessage
      - No shared memory = no races
      - The "JavaScript way"

4. When do you need synchronization?
   - Worker threads with SharedArrayBuffer? YES
   - Regular async/await? NO
   - Promises? NO
   - setTimeout/setInterval? NO

5. Best practices:
   - Default: Use async/await (no locks needed)
   - Worker threads: Prefer message passing over shared memory
   - If shared memory: Use Atomics
   - Don't create worker threads unless truly needed

6. Why JavaScript is different:
   - Python/Java/Go: Multi-threaded by default
   - JavaScript: Single-threaded with opt-in parallelism
   - Most JS code never needs locks
   - But when using workers, same rules apply!

⚠️  Don't assume "JavaScript is safe from races" - Worker threads +
    SharedArrayBuffer have the SAME race conditions as other languages!

💡 Run this example:
    node page_view_counter.js
`);
}

// Run the main function
main().catch(console.error);

File: low-level-design/lld-coding/multi-language/04-page-view-counter/page_view_counter.py
"""
Python: Page View Counter - Race Condition Demo
================================================
Shows:
1. Broken version (race condition)
2. Fixed version (with Lock)
3. Why Python's GIL doesn't save you
"""

import threading
import time


print("=" * 80)
print("          PYTHON: RACE CONDITION DEMONSTRATION")
print("=" * 80)


# ============================================================================
# Version 1: BROKEN - Race Condition
# ============================================================================
class PageViewCounterBroken:
    def __init__(self):
        self.views = 0

    def increment(self):
        # This looks atomic but it's NOT!
        # Under the hood: READ, ADD, WRITE (3 operations)
        temp = self.views
        time.sleep(0.0001)  # Simulate work, increase race window
        self.views = temp + 1


print("\n1. BROKEN VERSION (No Lock)")
print("-" * 80)

counter_broken = PageViewCounterBroken()
threads = []

# Create 1000 threads, each incrementing once
for i in range(1000):
    t = threading.Thread(target=counter_broken.increment)
    threads.append(t)
    t.start()

# Wait for all threads
for t in threads:
    t.join()

print(f"Expected: 1000")
print(f"Actual:   {counter_broken.views}")
print(f"Lost:     {1000 - counter_broken.views} increments")
print(f"Status:   {'✅ Correct' if counter_broken.views == 1000 else '❌ RACE CONDITION DETECTED!'}")


# ============================================================================
# Version 2: FIXED - Using Lock
# ============================================================================
class PageViewCounterFixed:
    def __init__(self):
        self.views = 0
        self.lock = threading.Lock()  # Create a lock

    def increment(self):
        with self.lock:  # Acquire lock (only one thread at a time)
            temp = self.views
            time.sleep(0.0001)  # Even with delay, no race!
            self.views = temp + 1
        # Lock automatically released here


print("\n2. FIXED VERSION (With Lock)")
print("-" * 80)

counter_fixed = PageViewCounterFixed()
threads = []

for i in range(1000):
    t = threading.Thread(target=counter_fixed.increment)
    threads.append(t)
    t.start()

for t in threads:
    t.join()

print(f"Expected: 1000")
print(f"Actual:   {counter_fixed.views}")
print(f"Status:   {'✅ Correct!' if counter_fixed.views == 1000 else '❌ Still broken?!'}")


# ============================================================================
# Version 3: PRODUCTION - Context Manager Pattern
# ============================================================================
class PageViewCounterProduction:
    """Production-ready version with proper encapsulation"""

    def __init__(self):
        self._views = 0  # Private
        self._lock = threading.Lock()

    def increment(self):
        """Thread-safe increment"""
        with self._lock:
            self._views += 1

    def get_views(self):
        """Thread-safe read"""
        with self._lock:
            return self._views

    def reset(self):
        """Thread-safe reset"""
        with self._lock:
            self._views = 0


print("\n3. PRODUCTION VERSION (Best Practices)")
print("-" * 80)

counter_prod = PageViewCounterProduction()


def simulate_user():
    """Simulate a user viewing pages"""
    for _ in range(10):
        counter_prod.increment()
        time.sleep(0.0001)


# Simulate 100 users, each viewing 10 pages
threads = []
for i in range(100):
    t = threading.Thread(target=simulate_user)
    threads.append(t)
    t.start()

for t in threads:
    t.join()

print(f"Expected: 1000 (100 users × 10 views)")
print(f"Actual:   {counter_prod.get_views()}")
print(f"Status:   ✅ Production-ready and correct!")


# ============================================================================
# Summary
# ============================================================================
print("\n" + "=" * 80)
print("                          SUMMARY")
print("=" * 80)
print("""
🎯 Key Lessons:

1. Python's GIL doesn't prevent all race conditions
   - GIL only prevents simultaneous bytecode execution
   - But threads can still interleave operations
   - ALWAYS use locks for shared data

2. The race happens between READ and WRITE
   - self.views = self.views + 1  # Looks atomic, but isn't!
   - Actually: READ → ADD → WRITE (3 steps)

3. Always use threading.Lock()
   - with lock:  # Pythonic and safe
   - Automatic acquire/release
   - Use for ALL shared mutable data

4. Best practices:
   - Make data private (_variable)
   - Provide thread-safe methods
   - Lock at method level, not caller level
   - Keep critical sections small

⚠️  In production, even a 0.1% failure rate = thousands of bugs!
""")
